src/consume/mod.rs

```diff
@@ -0,0 +1,202 @@
+//! Types, traits, and functions necessary to consume messages using hedwig
+//!
+//! See the [`Consumer`] trait.
+
+use crate::ValidatedMessage;
+use async_trait::async_trait;
+use either::Either;
+use futures_util::stream;
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Message consumers ingest messages from a queue service and present them to the user application
+/// as a [`Stream`](futures_util::stream::Stream).
+///
+/// ## Acknowledging Messages
+/// Typically message services deliver messages with a particular delivery time window, during
+/// which this message won't be sent to other consumers. In AWS SQS this is called the [visibility
+/// timeout][AWS], and in GCP PubSub this is the [ack deadline][GCP].
+///
+/// If a message is successfully acknowledged within this time, it will be considered processed and
+/// not delivered to other consumers (and possibly deleted depending on the service's
+/// configuration). A message can conversely be negatively-acknowledged, to indicate e.g.
+/// processing has failed and the message should be delivered again to some consumer. This time
+/// window can also be modified for each message, to allow for longer or shorter message processing
+/// than the default configured time window.
+///
+/// Implementations of this trait do not ack/nack/modify messages themselves, and instead present
+/// this functionality to users with the [`AcknowledgeableMessage`] type. Message processors are
+/// responsible for handling message acknowledgement, including extensions for processing time as
+/// necessary.
+// If we had async drop, sending nacks on drop would be nice. Alas, rust isn't there yet
+///
+/// Bear in mind that message delivery and acknowledgement are all best-effort in distributed
+/// message services. An acknowledged or extended message may still be re-delivered for any number
+/// of reasons, and applications should be made resilient to such events.
+///
+/// [AWS]: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html
+/// [GCP]: https://cloud.google.com/pubsub/docs/subscriber
+pub trait Consumer {
+    /// The type of acknowledgement tokens produced by the underlying service implementation
+    type AckToken: AcknowledgeToken;
```

Comment:
Nit: abbreviation here seems odd, given that it has a bound of trait with a full name.

Reply:
It felt like an `Iter` vs `Iterator` kind of thing. Concrete types would be abbreviated, traits would not